check/tests/fail.rs

```diff
@@ -0,0 +1,280 @@
+    let _ = ::env_logger::init();
+    let text = r#"
+let { y } = { x = 1 }
+2
+"#;
+    let result = typecheck(text);
+    assert_err!(result, UndefinedField(..));
+}
+
+#[test]
+fn duplicate_type_definition() {
+    let _ = ::env_logger::init();
+    let text = r#"
+type Test = Int
+in
+type Test = Float
+in 1
+"#;
+    let result = typecheck(text);
+    assert_err!(result, DuplicateTypeDefinition(..));
+}
+
+#[test]
+fn no_matching_overloaded_binding() {
+    let _ = ::env_logger::init();
+    let text = r#"
+let f x = x #Int+ 1
+in
+let f x = x #Float+ 1.0
+in f ""
+"#;
+    let result = typecheck(text);
+    assert_err!(result, Rename(..));
+}
+
+#[test]
+fn no_matching_binop_binding() {
+    let _ = ::env_logger::init();
+    let text = r#"
+let (++) x y = x #Int+ y
+let (++) x y = x #Float+ y
+"" ++ ""
+"#;
+    let result = typecheck(text);
+    assert_err!(result, Rename(..));
+}
+
+#[test]
+fn not_enough_information_to_decide_overload() {
+    let _ = ::env_logger::init();
+    let text = r#"
+let f x = x #Int+ 1
+let f x = x #Float+ 1.0
+\x -> f x
+"#;
+    let result = typecheck(text);
+    assert_err!(result, Rename(..));
+}
+
+#[test]
+fn type_field_mismatch() {
+    let _ = ::env_logger::init();
+    let text = r#"
+if True then
+    type Test = Int
+    { Test }
+else
+    type Test = Float
+    { Test }
+"#;
+    let result = typecheck(text);
+    assert_unify_err!(result, TypeMismatch(..));
+}
+
+#[test]
+fn arguments_need_to_be_instantiated_before_any_access() {
+    let _ = ::env_logger::init();
+    // test_fn: forall a. (a -> ()) -> ()
+    // To allow any type to be passed to `f` it should be
+    // test_fn: (forall a. a -> ()) -> ()
+    let text = r#"
+let test_fn f: (a -> ()) -> () =
+    f 2.0
+1
+"#;
+    let result = typecheck(text);
+    assert_unify_err!(result, TypeMismatch(..));
+}
+
+#[test]
+fn infer_ord_int() {
+    let _ = ::env_logger::init();
+    let text = r#"
+type Ordering = | LT | EQ | GT
+type Ord a = {
+    compare : a -> a -> Ordering
+}
+let ord_Int = {
+    compare = \l r ->
+        if l #Int< r
+        then LT
+        else if l #Int== r
+        then EQ
+        else GT
+}
+let make_Ord ord =
+    let compare = ord.compare
+    in {
+        (<=) = \l r ->
+            match compare l r with
+                | LT -> True
+                | EQ -> True
+                | GT -> False
+    }
+let (<=) = (make_Ord ord_Int).(<=)
+
+"" <= ""
+"#;
+    let result = typecheck(text);
+    assert_unify_err!(result, TypeMismatch(..), TypeMismatch(..));
+}
+
+#[test]
+fn recursive_types_with_differing_aliases() {
+    let _ = ::env_logger::init();
+    let text = r"
+type Option a = | None | Some a
+type R1 = Option R1
+and R2 = Option R2
+
+let x: R1 = None
+let y: R2 = x
+y
+";
+    let result = typecheck(text);
+    assert_unify_err!(result, Other(SelfRecursive(..)));
+}
+
+#[test]
+fn detect_self_recursive_aliases() {
+    let _ = ::env_logger::init();
+    let text = r"
+type A a = A a
```

Comment:
Just clarifying: so this an alias, not a union of one variant, because it isn't preceded by a pipe?

Also, does this mean that types aren't loaded into the context top-down?